GitHub Actions blog post (#25)
model-checking/kani-verifier-blog: _posts/2022-12-21-easily-verify-your-rust-in-ci-with-kani.md

---
layout: post
title: "Easily verify your Rust in CI with Kani and Github Actions"
---

> Formal verification tools like Kani definitively check that certain classes of bugs will not occur in your program under any circumstance.

Reviewer: remove "definitively"

Author: done

> Kani uses mathematical techniques to explore your code over all inputs, meaning that while testing is a great way to catch the *first* bug, formally verifying your code with Kani guarantees that you've caught the *last* bug.

Reviewer: This is a very strong statement. Even because we don't catch every source of UB. So I would prefer if we tune this down.

Author: Toned down

To learn more about Kani, and how you can use it to prevent bugs in your Rust code, you can [read the Kani documentation](https://model-checking.github.io/kani/), or [consult the real-world examples on our blog]({{site.baseurl | prepend: site.url}}).

Code, however, is seldom static.
As code evolves, how can we ensure that it remains correct; i.e. how do we catch the *next* bug?
Many codebases already contain test suites and other checks that can run automatically.
Developers can use a platform like [GitHub Actions](https://github.com/features/actions) to run these tests upon every code change, as part of continuous integration, and even [ensure that the tests pass before the code can be merged](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging).

> This assures developers that a code change does not introduce a bug for all the inputs that the test suite exercises.

Reviewer: I wouldn't use `all the inputs`. Maybe I would say that `this assures developers that a code change does not introduce a bug for inputs covered by the test suite.`

Author: updated

> For example, on our own Kani repo, a CI status report on a PR might look like this:

Reviewer: TBH, I don't think you need this example.

Author: Removed

To get the latest version, visit the [Kani Verifier action](https://github.com/marketplace/actions/kani-rust-verifier) on the marketplace, and click on the green "Use Latest Version" button, which will give you a yaml snippet you can paste into your `.yml` file.
<img src="{{site.baseurl | prepend: site.url}}/assets/images/kani-verifier-action.png" alt="A list of tests and checks run on a pull request as part of continuous integration, with all checks required and passing" />

For example, for Kani 0.17, your `kani-ci.yml` file might look like this:

```yaml
name: Kani CI
on:
  pull_request:
  push:
jobs:
  run-kani:
    runs-on: ubuntu-20.04
    steps:
      - name: 'Checkout your code.'
        uses: actions/checkout@v3

# You can get the latest version from
# https://github.com/marketplace/actions/kani-rust-verifier
      - name: 'Run Kani on your code.'
        uses: model-checking/kani-github-action@v0.17
```

For more advanced use cases, we provide facilities to override the working directory, as well as to configure the Kani command itself.

> For example, the [s2n-quic](https://github.com/aws/s2n-quic) project uses [the following CI configuration](https://github.com/aws/s2n-quic/blob/main/.github/workflows/ci.yml#L613), which overrides the working directory, and enables the `--tests` option of `cargo kani`.

Reviewer: This is a nice example.

Reviewer 2: We probably want to link to a hash rather than `main` since this line can change and it would be nice for the blog link to stay correct.

https://github.com/aws/s2n-quic/blob/c221530beb37addf81c7ed58461a945750a1b251/.github/workflows/ci.yml#L613

---
layout: post

> title: "Easily verify your Rust in CI with Kani and Github Actions"

Reviewer 3: Just want to say: I think this is a good title.

Author: 👍

Developers can use a platform like [GitHub Actions](https://github.com/features/actions) to run these tests upon every code change, as part of continuous integration, and even [ensure that the tests pass before the code can be merged](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging).
This assures developers that a code change does not introduce a bug for all the inputs that the test suite exercises.
For example, on our own Kani repo, a CI status report on a PR might look like this:

<img src="{{site.baseurl | prepend: site.url}}/assets/images/kani-ci-checks.png" alt="A list of tests and checks run on a pull request as part of continuous integration, with all checks required and passing" />

For developers looking for a higher assurance of software quality, Kani proofs can be validated automatically as part of continuous integration.
To this end, we are excited to announce the [Kani Rust Verifier Action](https://github.com/marketplace/actions/kani-rust-verifier) on the GitHub Marketplace.

## How to use the Kani GitHub action

> If you have a [Rust Cargo](https://doc.rust-lang.org/cargo/) project in a GitHub repo, using the Kani CI action is as simple as adding it to your `ci.yaml` file.

Reviewer 3: We might want to say "workflow file" or "github actions workflow file" instead since `ci` won't be universal?

Author: done

Your Kani checks will appear as part of the same workflow as your existing tests, as shown in this [s2n-quic](https://github.com/aws/s2n-quic) CI report:

<img src="{{site.baseurl | prepend: site.url}}/assets/images/kani-in-s2n-quic-ci.png" alt="s2n-quic CI report showing both tests and Kani proofs verified on a pull request" />

> To get the latest version, visit the [kani verifier action](https://github.com/marketplace/actions/kani-rust-verifier) on the marketplace, and click on the green “Use Latest Version” button, which will give you a yaml snippet you can paste into your `.yml` file.

Reviewer 3: Oh sweet! This is awesome to know. TIL :)

        uses: actions/checkout@v3

# You can get the latest version from
# https://github.com/marketplace/actions/kani-rust-verifier
      - name: 'Run Kani on your code.'
        uses: model-checking/kani-github-action@v0.17
```

For more advanced use cases, we provide facilities to override the working directory, as well as to configure the Kani command itself.
For example, the [s2n-quic](https://github.com/aws/s2n-quic) project uses [the following CI configuration](https://github.com/aws/s2n-quic/blob/main/.github/workflows/ci.yml#L613), which overrides the working directory, and enables the `--tests` option of `cargo kani`.

```yaml
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
# ...
jobs:
  kani:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
          submodules: true

      - name: Kani run
        uses: model-checking/kani-github-action@v0.17
        with:
          working-directory: quic/s2n-quic-core
          args: --tests
```

For full details, consult [the documentation](https://model-checking.github.io/kani/install-github-ci.html).

## When to run the Kani action

> CI systems such as GitHub Actions offer multiple options CI workflows, such as the Kani action, should execute.

Reviewer 3: Wording seems broken here

Author: fixed

---
layout: post
title: "Easily verify your Rust in CI with Kani and Github Actions"
---

> Formal verification tools, such as Kani, are valuable because they provide a way to definitively check, using mathematical techniques, whether a property of your code is true under all circumstances.

Reviewer 4: I wonder if the first sentence would be punchier if broken up, there's a lot to unpack here at the moment.

I also think that there is an implicit leap from "property of your code" to "catching bugs". It might be obvious to us that the 'property' here is bug-freedom of some sort, but I think this is jargon. Can I suggest:
```suggestion
Formal verification tools like Kani definitively check that certain classes of bugs will not occur in your program under any circumstance. Kani uses mathematical techniques to explore your code over all inputs, meaning that while testing is a great way to catch the *first* bug, formally verifying your code with Kani guarantees that you've caught the *last* bug.
```

To put it colloquially: testing is great for catching the **first** bug, formal verification is great for knowing you’ve caught the **last** bug.
To learn more about Kani, and how you can use it to prevent bugs in your Rust code, you can [read the Kani documentation](https://model-checking.github.io/kani/), or [consult the real-world examples on our blog](../../..).

> Code, however, is seldom static.

Reviewer 4: > by ensuring that tests are re-run on all code-changes, CI

> provide facilities to automatically incorporate CI...run checks, ranging from tests, to linters...

These sentences seem to imply that CI is a service/tool for running tests upon code changes. [Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration) is much broader than that, it is a *process* of continuously integrating code into a central repo, and having an automated testing setup is just one small aspect of it.

Mentioning this because "Continuous Verification" below seems to be building on that analogy, and I'm not sure that it works. I think that running the proofs is part of CI, just like running the tests is.
```suggestion
Code, however, is seldom static. As code evolves, how can we ensure that it remains correct; i.e. how do we catch the *next* bug? Many codebases already contain test suites and other checks that can run automatically. Developers can use a platform like [GitHub Actions](https://github.com/features/actions) to run these tests upon every code change, as part of continuous integration, and even [ensure that the tests pass before the code can be merged](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging). This assures developers that a code change does not introduce a bug for all the inputs that the test suite exercises.

For developers looking for a higher assurance of software quality, Kani proofs can be validated automatically as part of continuous integration. To this end, we are excited to announce the[Kani Rust Verifier Action](https://github.com/marketplace/actions/kani-rust-verifier) on the GitHub Marketplace.
```

As code evolves, how can we ensure that it remains correct; i.e.
how do we catch the **next** bug?
In the testing world, this is where **Continuous Integration** (CI) shines: by ensuring that tests are re-run on all code-changes, CI enables developers to catch bugs before they impact users.
Many platforms, such as GitHub, provide facilities to automatically incorporate CI into existing development workflows.
For example, [GitHub Actions](https://github.com/features/actions) provides developers with a facility to run checks, ranging from tests, to linters, to code-formatters, and have the results displayed on the pull request (PR) in real time.
Repositories can even [require that certain checks pass before merging a PR](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging).
For example, on our own Kani repo, a CI status report on a PR might look like this:

> <img src="{{site.baseurl | prepend: site.url}}/assets/images/kani-ci-checks.png" alt="Kani CI Checks" />

Reviewer 4: ```alt="A list of tests and checks run on a pull request as part of continuous integration, with all checks required and passing" />```

> Similarly, **Continuous Verification** enables developers to prevent regressions before they happen.

Reviewer 4: See comment above. Apart from that, I'm not sure that emphasizing "continuous verification" new term is warranted. It's not used anywhere else in the blog post, and I don't think we're particularly advocating this term (or are we?), so it seems distracting to introduce a new term

Kani users have asked us for an easy way to integrate continuous verification into their existing development workflows.
To this end, we are excited to announce the [Kani Rust Verifier Action](https://github.com/marketplace/actions/kani-rust-verifier) on the GitHub Marketplace.

## How to use the Kani GitHub action

If you have a [Rust Cargo](https://doc.rust-lang.org/cargo/) project in a GitHub repo, using the Kani CI action is as simple as adding it to your `ci.yaml` file.

To get the latest version, visit the [kani verifier action](https://github.com/marketplace/actions/kani-rust-verifier) on the marketplace, and click on the green “Use Latest Version” button, which will give you a yaml snippet you can paste into your `.yml` file.
<img src="{{site.baseurl | prepend: site.url}}/assets/images/kani-verifier-action.png" alt="Kani Verifier Action on GitHub Marketplace" />

> For example, for Kani 0.16, your `kani-ci.yml` file might look like this:

Reviewer 5: Would we want to have in here the most recent version of Kani when this goes live, which I believe is 0.17?

Reviewer 5: (this would also have to be updated in the code snippet below)

Author: fixed

```yaml
name: Kani CI
on:
  pull_request:
  push:
jobs:
  run-kani:
    runs-on: ubuntu-20.04
    steps:
      - name: 'Checkout your code.'
        uses: actions/checkout@v3

# You can get the latest version from
# https://github.com/marketplace/actions/kani-rust-verifier
      - name: 'Run Kani on your code.'
        uses: model-checking/kani-github-action@v0.16
```

For more advanced use cases, we provide facilities to override the working directory, as well as to configure the Kani command itself.

> For example, the [s2n-quic](https://github.com/aws/s2n-quic) project uses [the following CI configuration](https://github.com/aws/s2n-quic/blob/main/.github/workflows/ci.yml#L613), which overrides the working directory, and enables the `--tests` option of `cargo kani`.

Reviewer 4: Is there a good reason to use a screenshot of a text file? Having the actual text allows visually impaired readers to read.

Author: changed

<img src="{{site.baseurl | prepend: site.url}}/assets/images/s2n-quic-using-kani-action.png" alt="How s2n quic integrates the Kani action" />

For full details, consult [the documentation](https://model-checking.github.io/kani/install-github-ci.html).

## When to run the Kani action

> One question we have received from our users is “When should we run the Kani action?”  Is it better to run it on every PR, once on a PR right before merge, or nightly?  The answer is an emphatic “it depends!”.

Reviewer 4: Small fixes:
```suggestion
One question we have received from our users is “when should we run the Kani action?”  Is it better to run it on every PR, once on a PR right before merge, or nightly?  The answer is an emphatic “it depends!”.  The great thing about GitHub Actions is that you can use them in the way that best fits your workflow.  Our experience is that the sooner verification runs, the better: it’s easier to debug and fix issues when they’re found right away, rather than after they’re merged.  Our experience is that many Kani proofs complete quickly on the standard GitHub Action Runner.  Proofs that run in a similar time-frame to existing tests make sense to run on a similar cadence, e.g. on every pull request or code-push.
```

Reviewer 4: TBH I found this paragraph a bit more stilted than previous ones. I would avoid fluffy language like "the great thing about GitHub Actions", and continue writing in the prescriptive style you used in the rest of the post, rather than a conversation with 'users'...

Reviewer 4: > run it on every PR, once on a PR right before merge, or

I don't understand the difference. Do you mean "run it on every push to a PR"?

Author: Yes. Updated the language

The great thing about GitHub Actions are that you can use them in the way that best fits your workflow.
Our experience is that the sooner verification runs, the better: it’s easier to debug and fix issues when they’re found right away, rather than after they’re merged.

> Our experience is that many Kani proofs can complete quickly on standard CI hardware.

Reviewer 5: It's a little repetitive to begin two sentences in a row with "Our experience is that". Maybe the second sentence can be simplified to "Many Kani proofs complete quickly on standard CI hardware."

Author: Fixed

Proofs that run in a similar time-frame to existing tests make sense to run on a similar cadence, e.g.
on every pull request or code-push.

Some proofs, on the other hand, may be memory and compute intensive.
In these cases, your organization may prefer nightly jobs to reduce cost and avoid introducing latency to the PR process.
Our suggestion is to start with running proofs on every pull request, but start moving some of them to a separate nightly run if they begin to take too much time.

## Caveats 

> The Kani GitHub action executes on standard GitHub Action runners.

Reviewer 4: ```suggestion
The Kani GitHub Action executes on standard GitHub Action Runners; these machines [provide 7GB of RAM and 2 processing cores](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources).  Formal verification can be memory and compute intensive: if you find the Kani Action unning out of memory or timing out, we suggest using GitHub's [Large Runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners) or [AWS CodeBuild](https://aws.amazon.com/codebuild/).
```

Author: done

> By default, these machines are [not particularly powerful](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources): (7GB of RAM, 2 core processors).

Reviewer 4: ```suggestion
The Kani GitHub Action executes on standard GitHub Action Runners; these machines [provide 7GB of RAM and 2 processing cores](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources).
```
